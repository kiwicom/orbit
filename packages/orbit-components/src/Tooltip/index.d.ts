// Type definitions for @kiwicom/orbit-components
// Project: http://github.com/kiwicom/orbit

import * as React from "react";

import { Popper } from "../primitives/TooltipPrimitive";
import * as Common from "../common/common";

type Size = "small" | "medium";
export interface Props extends Common.Globals, Popper {
  readonly children: React.ReactNode;
  readonly content: React.ReactNode;
  readonly size?: Size;
  readonly renderInPortal?: boolean;
  readonly stopPropagation?: boolean;
  readonly enabled?: boolean;
  readonly tabIndex?: string | number;
  readonly removeUnderlinedText?: boolean;
  readonly block?: boolean;
  readonly lockScrolling?: boolean;
}

declare const Tooltip: React.FunctionComponent<Props>;
export { Tooltip, Tooltip as default };
