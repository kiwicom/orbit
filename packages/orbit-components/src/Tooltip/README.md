# Tooltip

To implement Tooltip component into your project you'll need to add the import:

```jsx
import Tooltip from "@kiwicom/orbit-components/lib/Tooltip";
```

After adding import into your project you can use it simply like:

```jsx
<Tooltip content="Your content">
  <Airplane />
</Tooltip>
```

## Props

Table below contains all types of the props available in the Tooltip component.

| Name                 | Type                      | Default  | Description                                                                                                                                      |
| :------------------- | :------------------------ | :------- | :----------------------------------------------------------------------------------------------------------------------------------------------- |
| **children**         | `React.Node`              |          | The reference element where the Tooltip will appear.                                                                                             |
| **content**          | `React.Node`              |          | The content to display in the Tooltip.                                                                                                           |
| dataTest             | `string`                  |          | Optional prop for testing purposes.                                                                                                              |
| id                   | `string`                  |          | Set `id` for `Tooltip`                                                                                                                           |
| enabled              | `boolean`                 | `true`   | Enable render of tooltip                                                                                                                         |
| block                | `boolean`                 | `false`  | Whether the children wrapper is inline or block. Useful when children need to take up the entire container width.                                |
| removeUnderlinedText | `boolean`                 |          | Removes underline on child component, when underline is not desired.                                                                             |
| size                 | [`size`](#size)           |          | The maximum possible size of the Tooltip.                                                                                                        |
| stopPropagation      | `boolean`                 |          | If `true` the click event on children won't bubble. Useful when you use Tooltip inside another clickable element.                                |
| lockScrolling        | `boolean`                 | `true`   | Whether to prevent scrolling of the rest of the page while Tooltip is open on mobile. This is on by default to provide a better user experience. |
| renderInPortal       | `boolean`                 | `true`   | Optional prop, set it to `false` if you're rendering Tooltip inside a custom portal, defaults to `true`                                          |
| placement            | [`placement`](#placement) | `auto`   | 12 different placements to choose from                                                                                                           |
| noFlip               | `boolean`                 | `false`  | Turns off automatic flipping of the Tooltip when there is not enough space                                                                       |
| offset               | `[number, number]`        | `[0, 5]` | Set offset `[vertical, horizontal]` for Tooltip                                                                                                  |

| tabIndex | `string \| number` | `"0"` | Specifies the tab order of an element |

## enum

| size       |
| :--------- |
| `"small"`  |
| `"medium"` |

## placement

| Placement        |
| :--------------- |
| `"top"`          |
| `"right"`        |
| `"bottom"`       |
| `"left"`         |
| `"top-start"`    |
| `"top-end"`      |
| `"right-start"`  |
| `"right-end"`    |
| `"bottom-start"` |
| `"bottom-end"`   |
| `"left-start"`   |
| `"left-end"`     |
| `"auto"`         |
| `"auto-start"`   |
| `"auto-end"`     |
| `"bottom"`       |

## Functional specs

- Whenever event `onMouseEnter`, `onFocus` or `onClick` fires, the script inside this component will calculate possible positions that can be applied and the first possible will be applied.

- You can prefer one position and one align that will be used if possible, otherwise the default order in [`enum`](#enum) table will be used for both position and align.

- For mobile devices, the user needs to click on the children to open the Tooltip.

- The Tooltip component supports rendering of many different components inside its children. You can use combination of e.g. Text, Stack, List and custom image to show information to user about the CVV code. All rendered text inside the Tooltip should have white color on big devices (>= largeMobile) automatically:

```jsx
<Tooltip
  content={
    <Stack>
      <CustomImage />
      <Text>You can find the CVV in the right corner of your credit card.</Text>
      <List>
        <ListItem>Additional information</ListItem>
      </List>
    </Stack>
  }
>
  <InformationCircle />
</Tooltip>
```

- If you use Tooltip inside Alert, the Text should be automatically underlined with dotted line colored into the color of Alert.

- For ensuring that Tooltip component will work properly, add `div` with `id` set to `tooltips`. If not, the Tooltip will be rendered into the end of your DOM.
