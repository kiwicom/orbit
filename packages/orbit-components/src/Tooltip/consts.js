// @flow

export const SIZE_OPTIONS = {
  SMALL: "small",
  MEDIUM: "medium",
};
