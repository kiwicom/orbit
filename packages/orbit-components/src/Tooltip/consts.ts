export enum SIZE_OPTIONS {
  SMALL = "small",
  MEDIUM = "medium",
}
