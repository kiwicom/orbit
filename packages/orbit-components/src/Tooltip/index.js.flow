// @flow
/*
  DOCUMENTATION: https://orbit.kiwi/components/tooltip/
*/
import * as React from "react";

import type { Popper } from "../primitives/TooltipPrimitive/index.js.flow";
import type { Globals } from "../common/common.js.flow";

export type Sizes = "small" | "medium";

export type Size = {|
  size: Sizes,
|};

export type Props = {|
  ...Globals,
  +children?: React.Node,
  +content: React.Node,
  +size?: Sizes,
  +onShow?: () => void | Promise<void>,
  +enabled?: boolean,
  +labelClose?: React.Node,
  +renderInPortal?: boolean,
  +stopPropagation?: boolean,
  +tabIndex?: string | number,
  +removeUnderlinedText?: boolean,
  +block?: boolean,
  +lockScrolling?: boolean,
  ...Popper,
|};

declare export default React.ComponentType<Props>;
