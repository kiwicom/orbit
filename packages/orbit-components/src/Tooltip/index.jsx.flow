// @flow
/*
  DOCUMENTATION: https://orbit.kiwi/components/tooltip/
*/
import * as React from "react";

import type { Globals } from "../common/common.js.flow";

export type Dimensions = {|
  containerTop: number,
  containerLeft: number,
  containerHeight: number,
  containerWidth: number,
  tooltipHeight: number,
  tooltipWidth: number,
  contentHeight: number,
  windowWidth: number,
  windowHeight: number,
  containerTopPure: number,
  containerLeftPure: number,
|};

export type Aligns = "center" | "start" | "end";

export type Align = {|
  align: Aligns,
|};

export type Positions = "right" | "left" | "top" | "bottom";

export type Position = {|
  position: Positions,
|};

export type Sizes = "small" | "medium";

export type Size = {|
  size: Sizes,
|};

export type Props = {|
  ...Globals,
  +children?: React.Node,
  +content: React.Node,
  +size?: Sizes,
  +renderInPortal?: boolean,
  +stopPropagation?: boolean,
  +preferredPosition?: Positions,
  +preferredAlign?: Aligns,
  +enabled?: boolean,
  +tabIndex?: string | number,
  +removeUnderlinedText?: boolean,
  +block?: boolean,
  +lockScrolling?: boolean,
|};

declare export default React.ComponentType<Props>;
