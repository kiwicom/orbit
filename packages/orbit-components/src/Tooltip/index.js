// @flow
import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";

import type { Props } from "./index";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  preferredPosition,
  preferredAlign,
  stopPropagation = false,
  removeUnderlinedText,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      preferredPosition={preferredPosition}
      preferredAlign={preferredAlign}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
