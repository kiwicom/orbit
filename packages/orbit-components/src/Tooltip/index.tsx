import * as React from "react";

import { SIZE_OPTIONS } from "./consts";
import TooltipPrimitive from "../primitives/TooltipPrimitive";
import useMediaQuery from "../hooks/useMediaQuery";
import MobileDialog from "../primitives/MobileDialogPrimitive";
import { Props } from "./types";

const Tooltip = ({
  children,
  enabled = true,
  tabIndex = "0",
  dataTest,
  size = SIZE_OPTIONS.SMALL,
  content,
  id,
  placement,
  lockScrolling,
  renderInPortal = true,
  stopPropagation = false,
  removeUnderlinedText,
  block = false,
}: Props) => {
  const { isLargeMobile } = useMediaQuery();
  return isLargeMobile ? (
    <TooltipPrimitive
      dataTest={dataTest}
      id={id}
      tabIndex={tabIndex}
      enabled={enabled}
      content={content}
      size={size}
      renderInPortal={renderInPortal}
      placement={placement}
      stopPropagation={stopPropagation}
      removeUnderlinedText={removeUnderlinedText}
      block={block}
    >
      {children}
    </TooltipPrimitive>
  ) : (
    <MobileDialog
      dataTest={dataTest}
      tabIndex={tabIndex}
      id={id}
      enabled={enabled}
      lockScrolling={lockScrolling}
      content={content}
      removeUnderlinedText={removeUnderlinedText}
      stopPropagation={stopPropagation}
      block={block}
    >
      {children}
    </MobileDialog>
  );
};

export default Tooltip;
