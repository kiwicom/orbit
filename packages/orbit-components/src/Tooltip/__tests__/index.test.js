// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Tooltip from "../index";
import Airplane from "../../icons/Airplane";

describe("Tooltip", () => {
  const content = "Write some message to the user";
  const component = shallow(
    <Tooltip content={content}>
      <Airplane />
    </Tooltip>,
  );
  it("it should match snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
