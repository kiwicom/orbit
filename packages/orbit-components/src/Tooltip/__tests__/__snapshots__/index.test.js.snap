// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip it should match snapshot 1`] = `
<Fragment>
  <Tooltip__StyledTooltipChildren
    aria-describedby={null}
    enabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    tabIndex="0"
  >
    <Airplane />
  </Tooltip__StyledTooltipChildren>
</Fragment>
`;
